metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -443,41 +597,165 @@ class PackageProvider(
+        toImport.partition(_.isGivenOrExtension)
+      val allToImport =
+        (toImportNotGivens.map(decl =>
+          (decl.innerPackageParts, decl.name)
+        ) ++ toImportGivens.map(decl => (decl.innerPackageParts, "given")))
+          .groupMap(_._1)(_._2)
+          .map { case (innerPackageParts, names) =>
+            (innerPackageParts :+ mkImportees(names.toList)).mkString(".")
+          }
+
+      if (newPackageName.isEmpty)
+        allToImport.map(sym => s"import $sym\n").mkString
+      else allToImport.map(sym => s"import $newPackageName.$sym\n").mkString
     }
-    go(tree).toList
-  }
 
-  private def findRefs(
-      name: Name,
-      uri: String,
-  ): List[(String, Location)] = {
-    val params = new ReferenceParams(
-      new TextDocumentIdentifier(uri),
-      new Position(name.pos.startLine, name.pos.startColumn),
-      new ReferenceContext(false), // do not include declaration
+    val optLastPackage = for {
+      tree <- optTree
+      lastPackage <- findPackages(tree).pkgs.lastOption
+    } yield lastPackage.pkg
+
+    val optLastImport = for {
+      lastPackage <- optLastPackage
+      lastImport <- lastPackage.stats.collect { case i: Import => i }.lastOption
+    } yield lastImport
+
+    val optNewImportRange = optLastImport
+      .map(_.pos)
+      .orElse(optLastPackage.map(_.ref.pos))
+      .map(posNextToEndOf(_))
+
+    val range = optNewImportRange.getOrElse(
+      new Range(new Position(0, 0), new Position(0, 0))
     )
-    refProvider.references(params).flatMap(s => s.locations.map((s.symbol, _)))
+    new TextEdit(range, importString)
   }
 
-  private def findPkgs(tree: Tree): PkgsStructure = {
+  private def findTopLevelDecls(
+      path: AbsolutePath,
+      topLevelPackageParts: List[String],
+  ): List[TopLevelDeclaration] = {
+    val filename = path.filename
+    val filenamePart = filename.stripSuffix(".scala").stripSuffix(".sc")
+    val content = new String(Files.readAllBytes(path.toNIO))
+    val input = Input.VirtualFile(filename, content)
+
+    def isPackageObjectLike(symbol: String) =
+      Set("package", filenamePart ++ "$package").contains(symbol)
+    def isTopLevelPackageOrPackageObject(symbol: String): Boolean =
+      symbol.isPackage || {
+        val (desc, owner) = DescriptorParser(symbol)
+        (isPackageObjectLike(desc.name.value)
+        && isTopLevelPackageOrPackageObject(owner))
+      }
+
+    def alternatives(si: s.SymbolInformation): Set[String] =
+      if (si.isCase) Set(s"${si.symbol.dropRight(1)}.") else Set.empty
+
+    val dialect = {
+      val optDialect =
+        for {
+          buidTargetId <- buildTargets.inverseSources(path)
+          scalaTarget <- buildTargets.scalaTarget(buidTargetId)
+        } yield ScalaVersions.dialectForScalaVersion(
+          scalaTarget.scalaVersion,
+          includeSource3 = true,
+        )
+      optDialect.getOrElse(dialects.Scala213)
+    }
+    m.internal.mtags.Mtags
+      .index(input, dialect)
```

Comment:
So I was hoping to use something like `onDemandSymbolIndexer` so we don't index things if it's already indexed, but it doesn't keep all the information I need (if symbol is extension/given) and all the maps there are the other way around.